pkg/provenance/provenance.go

```diff
@@ -147,122 +148,436 @@ func (o ObjectLineage) GetVersions() string {
+			fieldNum, err := strconv.Atoi(key[5:])
+			if err != nil {
+				return fmt.Sprintf("Failure, could not convert %s. Invalid Query parameters.", key)
+			}
+			//find associated value by looking in the map for value+fieldNum.
+			valueOfKey, ok := argMap["value"+strconv.Itoa(fieldNum)]
+			if !ok {
+				return fmt.Sprintf("Could not find an associated value for field: %s", key)
+			}
+			attributeValueSlice[0] = value
+			attributeValueSlice[1] = valueOfKey
+			allQueryPairs = append(allQueryPairs, attributeValueSlice)
 		}
 	}
-	return strconv.Itoa(-1)
+	// fmt.Printf("attributeValuePairs%s\n", allQueryPairs)
+
+	// This method is to build the attributeRelationships from the Query
+	// I Will use this to ensure that fields belonging to the same top-level
+	// attribute, will be treated as a joint query. So you can't just
+	// ask if username ever is daniel and password is ever 223843, because
+	// it could find that in different parts of the spec. They both must be satisfied in the same map object
+	mapRelationships := BuildAttributeRelationships(specs, allQueryPairs)
+	// fmt.Println(mapRelationships)
+	fmt.Println(specs)
+	for _, spec := range specs {
+
+		//every element represents whether a query pair was satisfied. they all must be true.
+		//if they all are true, then that will be the version where the query is first satisfied.
+		andGate := make([]bool, len(allQueryPairs))
+		index := 0
+		for _, pair := range allQueryPairs {
+			qkey := pair[0]
+			qval := pair[1]
+			satisfied := false
+			for mkey, mvalue := range spec.AttributeToData {
+				//search through the attributes in the spec. Possible types
+				//are string, array of strings, and a map. so I will need to check these
+				//with different search methods
+				vString, ok1 := mvalue.(string)
+				if ok1 { //if underlying value is a string
+					satisfied = handleTrivialFields(vString, qkey, qval, mkey)
+					if satisfied{
+						break //qkey/qval was satisfied somewhere in the spec attributes, so move on to the next qkey/qval
+					}
+				}
+
+				vStringSlice, ok2 := mvalue.([]string)
+				if ok2 { //if it is a slice of strings
+					satisfied = handleSimpleFields(vStringSlice, qkey, qval, mkey)
+					if satisfied{
+						break //qkey/qval was satisfied somewhere in the spec attributes, so move on to the next qkey/qval
+					}
+				}
+
+				vSliceMap, ok3 := mvalue.([]map[string]string)
+				//if it is a map, will need to check the underlying data to see
+				//if the qkey/qval exists below the top level attribute. username exists
+				//in the map contained by the 'users' attribute in the spec object, for example.
+				if ok3 {
+					satisfied = handleCompositeFields(vSliceMap, mapRelationships, qkey, qval, mkey)
+					if satisfied{
+						break //qkey/qval was satisfied somewhere in the spec attributes, so move on to the next qkey/qval
+						 
+					}
+				}
+			}
+			andGate[index] = satisfied
+			index += 1
```

Comment:
Do we really need to maintain/manage index?
Can we not simply use append?
andGate = append(andGate, satisfied)